ols/src/cache/redis_cache.py

```diff
@@ -70,14 +81,33 @@ def insert_or_append(self, user_id: str, conversation_id: str, value: str) -> No
             value: The value to set.
 
         Raises:
-            OutOfMemoryError: If item is evicted when Redis allocated
-                memory is higher than maxmemory.
+            RedisConnectionError: If item is unable to update after REDIS_MAX_RETRY.
         """
         key = super().construct_key(user_id, conversation_id)
-
         old_value = self.get(user_id, conversation_id)
-        with self._lock:
-            if old_value:
-                self.redis_client.set(key, old_value + "\n" + value)
-            else:
-                self.redis_client.set(key, value)
+        retry_count = 0
+        while retry_count < constants.REDIS_MAX_RETRY:
+            logger.debug("Updating redis cache ")
+            with self._lock:
```

Comment:
this is too broad a lock, to do this properly you should lock on the conversation id

Reply:
also you're locking after you retrieved the value from redis.....  which means you're not avoiding race conditions because another process could insert/update the value after you do your initial get.

And finally, as soon as we have multiple instances of OLS running, talking to the same shared redis cache, this local locking won't accomplish anything.  We need to be locking at the redis level, not the client level, if we want to ensure no race conditions.